src/wasm-interpreter.h

```diff
@@ -2228,6 +2235,119 @@ class ModuleInstance
     : ModuleInstanceBase(wasm, externalInterface) {}
 };
 
+// Evaluates an expression given its surrounding context. Errors if we hit
+// anything that can't be evaluated down to a constant.
+class ContextAwareExpressionRunner final
+  : public ExpressionRunner<ContextAwareExpressionRunner> {
+
+public:
+  // Map of `local.get`s to their respective values.
+  typedef std::unordered_map<LocalGet*, Literals> GetValues;
+
+  // Whether we are trying to precompute down to an expression (which we can
+  // do on say 5 + 6) or to a value (which we can't do on a local.tee that
+  // flows a 7 through it). When we want to replace the expression, we can
+  // only do so when it has no side effects. When we don't care about
+  // replacing the expression, we just want to know if it will contain a known
+  // constant.
+  enum FlagValues {
+    // Just evaluate the expression by default, where we can ignore some side
+    // effects like those of a `local.tee`, but not others like traps.
+    DEFAULT = 0,
+    // We are going to replace the expression afterwards, so side effects
+    // including those of `local.tee`s for example must be retained.
+    REPLACE = 1 << 0,
+    // We are going to execute the runner in a function-parallel scenario, so we
+    // cannot perform traversal of nodes that might become modified
+    // non-deterministically, like function calls, where the called function
+    // might or might not have been optimized already to something we can
+    // traverse successfully.
+    PARALLEL = 1 << 1
```

Comment:
I think `PARALLEL` is not obvious enough. How about `TRAVERSE_OTHER_FUNCS` which directly says what is the underlying issue here?

Reply:
Renamed the flags to `PRESERVE_SIDEEFFECTS` and `TRAVERSE_CALLS`